### ISSUE TITLE
Container placement strategy is not honoured

### ISSUE THREAD

**Issue by shemau** [concourse/concourse#4068]
Concourse 5.3 does not honour

    container_placement_strategy: fewest-build-containers

When this strategy is used concourse does not balance the workload across containers based on fly workers output.

Whilst one worker may have 40 containers, concourse will still place both resource checks and task steps on a worker with 256 containers.
Whilst there are in excess of 1000 free container spaces, concourse will still place containers on a worker with 256 containers.

This behaviour makes concourse unscalable and unstable, in so much as it renders concourse unable to run tasks when system utilisation is around 30 to 40% or when one worker is restarted for any reason including an IaaS issue.

There are so many related issues, for #3251 which is closed apparently without resolution. How is #3288 a fix? Randomly place containers?

Is there any placement strategy that is effectively anywhere-that-is-not-full?

**Comment by ddadlani** [comment]
Hi @shemau, the bug you are looking for is likely #3301. `fewest-build-containers` fails to place correctly if several steps are kicked off at the same time, because there is no container scheduling lock.

#3288 was to place *check* containers randomly, not all containers. Since check containers are so lightweight, there is no correlation between high worker load and large numbers of check containers.

The fix for #3301 was merged into master a few days ago (i.e. after v5.3), but there are still some issues related to the fix (#4070) that are being worked out.

**Comment by shemau** [comment]
I will wait for a build with all those fixes.  #3301 has been reverted by #4070 and is reopened.

It seems an issue, not really covered above, that there is no mechanism to throttle work (specifically check container work) to a worker that is exceeding the garden threshold. There does seem to be a hard limit of 256, workers never report more than 256. 

Another issue, is that workers bounce on 256 tend to stay there and require a manual drain and restart. It is unclear if this is due to more work being undertaken, 'failed' containers being retained or check containers having a lifetime (1 hour?).
We see this during every update now. 15 workers with 3750 peak container capacity. At stable runtime there are in the order of 1300 containers. Following a concourse upgrade that cycles workers, there are in the order 2700 containers with maxed out workers and a creeping sickness that left unchecked stops all productive work. It becomes necessary to start manually draining, cleaning disks and restarting overloaded workers to manually rebalance the work. Each time a 256 worker is stopped there are minimal jobs relocated to other workers (total containers drops). Eventually stability is achieved and workload continues at the original 1300 containers level.

**Comment by ddadlani** [comment]
The mechanism to throttle work that you mention is being discussed in #3695

**Comment by shemau** [comment]
Is there any outlook on when a any moderately sensible container placement strategy will be available for both task and check containers?

Today six worker (overlooking the other 8) with  1, 9, 27, 207, 214 and 220 container counts. A few minutes later two of the works were at 227 and 231. Why doesn't 1 get ALL work until it reaches of exceeds 9, they both share the work until they reach 27 and then all three share the work?
It is appreciated containers per worker is an approximation. It is appreciated there might be a race condition resulting in a little bit too much work starting in parallel may end up in one place. Honestly 1 is nothing like 220, it is not sensible to just randomly spray jobs, even if they are possibly light weight, when there is a unchecked hard limit from garden.

I am unaware of any option that would provide any relief. Adding more workers is not a solution, we have already done that. Setting the placement strategy is not a solution, it is not honoured. 

Are there any options that could help here?

**Comment by ddadlani** [comment]
Hi @shemau,

The scheduling logic of Concourse is currently being reworked to address the very same issues that you mention. This is why I directed you to #3301 and #3695, because both those issues are discussing the things you are suggesting, and you can also keep track of our progress as we update those issues with our work. 

As a workaround you can use the `random` container placement strategy as the differences between containers on your workers should not be as drastic as they are now.

I suggest that we continue this discussion on one or both of those issue threads since they encapsulate a lot of context that you might find useful to know. 

Since this issue is essentially duplicating those two, I will close this for now. Feel free to open another issue if you feel like what you need does not belong under either of those two issues.
